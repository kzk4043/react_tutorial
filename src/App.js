import { useState } from "react";

export function Square({ value, onSquareClick, ...props }) {
  return (
    <button className="square" onClick={onSquareClick} {...props}>
      {value}
    </button>
  );
}

export function Board({ xIsNext, squares, onPlay }) {
  function handleClick(i) {
    if (squares[i] || calculateWinner(squares)) {
      return;
    }

    const nextSquares = squares.slice();
    nextSquares[i] = xIsNext ? "X" : "O";

    onPlay(nextSquares);
  }

  const winner = calculateWinner(squares);

  return (
    <>
      <div className="status">{calculateStatus(squares, winner, xIsNext)}</div>
      <div className="board-row">
        <Square value={squares[0]} onSquareClick={() => handleClick(0)} />
        <Square value={squares[1]} onSquareClick={() => handleClick(1)} />
        <Square value={squares[2]} onSquareClick={() => handleClick(2)} />
      </div>
      <div className="board-row">
        <Square value={squares[3]} onSquareClick={() => handleClick(3)} />
        <Square value={squares[4]} onSquareClick={() => handleClick(4)} />
        <Square value={squares[5]} onSquareClick={() => handleClick(5)} />
      </div>
      <div className="board-row">
        <Square value={squares[6]} onSquareClick={() => handleClick(6)} />
        <Square value={squares[7]} onSquareClick={() => handleClick(7)} />
        <Square value={squares[8]} onSquareClick={() => handleClick(8)} />
      </div>
    </>
  );
}

const calculateWinner = (squares) => {
  const lines = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6],
  ];

  for (const line of lines) {
    const [a, b, c] = line;
    if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
      return squares[a];
    }
  }
  return null;
};

const calculateStatus = (squares, winner, xIsNext) => {
  if (!winner && !squares.includes(null)) {
    return `Draw!`;
  }

  return winner ? `Winner: ${winner}` : `Next player: ${xIsNext ? "X" : "O"}`;
};

export default function Game() {
  const [history, setHistory] = useState([Array(9).fill(null)]);
  const [currentMove, setCurrentMove] = useState(0);
  const xIsNext = currentMove % 2 === 0;
  const currentSquares = history[currentMove];

  function handlePlay(nextSquares) {
    const nextHistory = [...history.slice(0, currentMove + 1), nextSquares];
    setHistory(nextHistory);
    setCurrentMove(nextHistory.length - 1);
  }

  function jumpTo(nextMove) {
    setCurrentMove(nextMove);
  }

  const moves = history.map((squares, move) => {
    const description = move > 0 ? `Go to move #${move}` : "Go to game start";

    return (
      <li key={move}>
        <button onClick={() => jumpTo(move)}>{description}</button>
      </li>
    );
  });

  return (
    <div className="game">
      <div className="game-board">
        <Board xIsNext={xIsNext} squares={currentSquares} onPlay={handlePlay} />
      </div>
      <div className="game-info">
        <ol>{moves}</ol>
      </div>
    </div>
  );
}
