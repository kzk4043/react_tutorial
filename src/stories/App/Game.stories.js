import Game from "../../App";
import { within, userEvent } from "@storybook/testing-library";
import { expect } from "@storybook/jest";

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
export default {
  title: "Tic-Tac-Toe/Game",
  component: Game,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
};

// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args
// Gameの初期状態
export const Initial = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const squares = canvas.getAllByRole("button");
    await userEvent.click(squares[0]);
    await userEvent.click(squares[1]);
    await userEvent.click(squares[2]);
    await userEvent.click(squares[3]);
    await userEvent.click(squares[4]);
    await userEvent.click(squares[5]);
    await userEvent.click(squares[6]);

    // 👇 Assert DOM structure
    await expect(canvas.getByText("Winner: X")).toBeInTheDocument();
  },
};

// 引き分けの時は「Draw!」と表示される
// 引き分けなので、勝利メッセージが表示されないことを確認
export const Draw = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const squares = canvas.getAllByRole("button");
    await userEvent.click(squares[0]);
    await userEvent.click(squares[1]);
    await userEvent.click(squares[2]);
    await userEvent.click(squares[3]);
    await userEvent.click(squares[5]);
    await userEvent.click(squares[4]);
    await userEvent.click(squares[6]);
    await userEvent.click(squares[8]);
    await userEvent.click(squares[7]);

    // 👇 Assert DOM structure
    await expect(canvas.getByText("Draw!")).toBeInTheDocument();
    await expect(canvas.queryByText("Winner")).not.toBeInTheDocument();
  },
};
