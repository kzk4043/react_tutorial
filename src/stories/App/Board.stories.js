import { Board } from "../../App";

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
export default {
  title: "Tic-Tac-Toe/Board",
  component: Board,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
};

// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args
// xIsNextがtrueの場合
export const XIsNext = {
  args: {
    xIsNext: true,
    squares: Array(9).fill(null),
    onPlay: () => {
      console.log("Square clicked");
    },
  },
};

// xIsNextがfalseの場合
export const OIsNext = {
  args: {
    xIsNext: false,
    squares: Array(9).fill(null),
    onPlay: () => {
      console.log("Square clicked");
    },
  },
};

// squaresの中身が全部Xの場合
export const X = {
  args: {
    xIsNext: true,
    squares: Array(9).fill("X"),
    onPlay: () => {
      console.log("Square clicked");
    },
  },
};

// squaresの中身が全部Oの場合
export const O = {
  args: {
    xIsNext: false,
    squares: Array(9).fill("O"),
    onPlay: () => {
      console.log("Square clicked");
    },
  },
};

// squaresの中身が全部△の場合
export const Triangle = {
  args: {
    xIsNext: true,
    squares: Array(9).fill("△"),
    onPlay: () => {
      console.log("Square clicked");
    },
  },
};

// Xが勝利した場合
export const XWin = {
  args: {
    xIsNext: true,
    squares: ["X", "X", "X", "O", "O", null, null, null, null],
    onPlay: () => {
      console.log("Square clicked");
    },
  },
};

// Oが勝利した場合
export const OWin = {
  args: {
    xIsNext: true,
    squares: ["O", "O", "O", "X", "X", "X", null, null, null],
    onPlay: () => {
      console.log("Square clicked");
    },
  },
};
