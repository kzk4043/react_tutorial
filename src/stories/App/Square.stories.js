import { Square } from "../../App";

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
export default {
  title: "Tic-Tac-Toe/Square",
  component: Square,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
  argTypes: {
    onClick: { action: "clicked" },
  },
};

// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args
export const Void = {
  args: {
    value: "",
    onSquareClick: () => {
      console.log("Square clicked");
    },
  },
};

// valueがXの場合
export const X = {
  args: {
    value: "X",
    onSquareClick: () => {
      console.log("Square clicked");
    },
  },
};

// valueがOの場合
export const O = {
  args: {
    value: "O",
    onSquareClick: () => {
      console.log("Square clicked");
    },
  },
};

// valueが△の場合
export const Triangle = {
  args: {
    value: "△",
    onSquareClick: () => {
      console.log("Square clicked");
    },
  },
};

// 課題とは関係ないが、parameters機能を使ってみる
export const CheckParameters = {
  args: {
    value: "",
    onSquareClick: () => {
      console.log("Square clicked");
    },
  },
  parameters: {
    backgrounds: {
      values: [
        { name: "red", value: "#f00" },
        { name: "green", value: "#0f0" },
        { name: "blue", value: "#00f" },
      ],
    },
  },
};

// 課題とは関係ないが、actions addonを使ってみる
export const CheckActions = {
  args: {
    value: "",
    onSquareClick: () => {
      console.log("Square clicked");
    },
  },
};
