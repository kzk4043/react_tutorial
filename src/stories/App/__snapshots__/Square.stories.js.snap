// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tic-Tac-Toe/Square CheckActions smoke-test 1`] = `
<button class="square">
</button>
`;

exports[`Tic-Tac-Toe/Square CheckParameters smoke-test 1`] = `
<button class="square">
</button>
`;

exports[`Tic-Tac-Toe/Square O smoke-test 1`] = `
<button class="square">
  O
</button>
`;

exports[`Tic-Tac-Toe/Square Triangle smoke-test 1`] = `
<button class="square">
  â–³
</button>
`;

exports[`Tic-Tac-Toe/Square Void smoke-test 1`] = `
<button class="square">
</button>
`;

exports[`Tic-Tac-Toe/Square X smoke-test 1`] = `
<button class="square">
  X
</button>
`;
